Feature parity for subsasgn and subsref

DionTimmermann opened the pull request:
Bugfixed and new features for subsasgn and subsref to make these functions behave the same way as they do with double varaibles in matlab. The tests used can be found in https://github.com/DionTimmermann/metas-unclib-matlab-wrapper-tests .

Practically complete rewrite of subsasgn:
- Fixed several bugs that were caused by the assumption that the indexes (S.subs) and B would always have the same shape/dimension.
- Removing these bugs also adds support for (partial) linear indexing
- Added support for logical indexes.
- Added support for null assignment (a(2:3) = [];).
- Added several error messages.
- Removed support for dot indexing, as this seemed to only result in internal errors.
- Some optimizations to prevent repeated calls of size/numel/isvector and unnecessary typecasts. With these optimizations the new implementation is as fast as the old, even though it tests for several errors.
With these changes, subsasgn now behaves the same for double and LinProp/DistProp/MCProp variables (except for some error messages in special cases).

Major update to subsref:
- Fixed a bug where the size of the matrix returned was not always as expected (special cases for linear indexing and the orientation of vectors).
- Added support for (partial) linear indexing.
- Added support for logical indexes
- Fixed a bug, where only one trailing singleton dimension was removed, instead of all.
- Added several error messages.
- Removed support for dot indexing, as this seemed to only result in internal errors.
- Some optimizations to prevent repeated calls of size/numel/isvector. With these optimizations the new implementation is as fast as the old, even though it tests for several errors.
With these changes, subsref now behaves the same for double and LinProp/DistProp/MCProp variables (except for some error messages in special cases).

wollmich commented on @LinProp/LinProp.m:
```diff
@@ -431,161 +431,347 @@ function display(obj)
+                else
+                    if dimI == 1
+                        if strcmp(I, ':')
+                            C = []; % TODO: Do we need to destroy/free something?
```
On the .NET side we don't need to destroy or free something. This is done by the garbage collector.

wollmich commented on @LinProp/LinProp.m:
```diff
@@ -431,161 +431,347 @@ function display(obj)
+        %   of that size.
+        
+            if strcmp('.', {S.type})
+                error('Dot indexing is not supported for variables of this type.');   %TODO: This was allowed in  V2.4.7 of the uncLib MATLAB wrapper - but did not serve any function, as far as Dion can see.
```
Dot indexing is not needed. And it was not working in V2.4.7.
So we can remove the TODO comment here.

wollmich commented on @MCProp/MCProp.m:
```diff
@@ -431,161 +431,343 @@ function display(obj)
+        %   the subscripts.  The result is LENGTH(I)-by-LENGTH(J)-by-LENGTH(K)-...
+        
+            if strcmp('.', {S.type})
+                error('Dot indexing is not supported for variables of this type.');   %TODO: This was allowed in  V2.4.7 of the uncLib MATLAB wrapper - but did not serve any function, as far as Dion can see.
```
Remove TODO comment. Dot indexing is not needed.

wollmich commented on @LinProp/LinProp.m:
```diff
@@ -431,161 +431,347 @@ function display(obj)
+        %   the subscripts.  The result is LENGTH(I)-by-LENGTH(J)-by-LENGTH(K)-...
+        
+            if strcmp('.', {S.type})
+                error('Dot indexing is not supported for variables of this type.');   %TODO: This was allowed in  V2.4.7 of the uncLib MATLAB wrapper - but did not serve any function, as far as Dion can see.
```
Dot indexing is not needed. And it was not working in V2.4.7.
So we can remove the TODO comment here.

wollmich commented on @LinProp/LinProp.m:
```diff
@@ -431,161 +431,347 @@ function display(obj)
+            end
+            
+            % TODO: We could probably simplify this code if we would used 
+            % SetItemsNd instead of SetItems1d for 1-by-N vectors.
```
SetItemsNd needs to compute the flatindex inside METAS UncLib. SetItems1d doesn't compute the flatindex.
The overhead of using SetItemsNd is not too much. Only integer multiplication and an additional array initialization.

So I think we should simplify the MATLAB wrapper.

DionTimmermann replied:
This idea proved to be more trouble than it was worth. Instead, I simplified the code be reodering code segements and replacing all uses of the variable transpose_vector with the variable output_shape.

wollmich reviewed (commented):
See my comments in the LinProp file.

The following two test are not working anymore with your proposed code:

```
clear g
g(5) = LinProp(1)

clear h
h(5) = LinProp(1i)
```

DionTimmermann commented:
Due to the line 
```
% direct assignment of a subpart of a matrix does not work
v3(2,:) = [x1 x2]; %DOES NOT WORK! RETURNS ERROR
```
in the Tutorial.m, I thought `clear g; g(5) = LinProp(1);` would also not be possible. Since it is possible, `clear g; g(:) = LinProp(5);` and `clear g; g(:, :) = LinProp(magic(3));` should also be possible. I added a patch that adds this behavior.

wollmich commented:
When the tutorial was written, `clear g; g(5) = LinProp(1);` was not possible. But it was needed if someone was using `parfor` for example. Thanks for adding the other cases with `g(:)` and `g(:, :)`.

But the following test is still failing: `clear g; g(5:8, :) = 1i.*LinProp(rand(4))`. If I use this it works `clear g; g(5:8, 1:4) = 1i.*LinProp(rand(4))`.

wollmich commented on @LinProp/LinProp.m:
```diff
@@ -431,161 +431,382 @@ function display(obj)
+                end
+            end
+
+            % Calucate size of output vector
```
Typo: Calculate

wollmich commented on @LinProp/LinProp.m:
```diff
@@ -431,161 +431,382 @@ function display(obj)
-                    if ((ni <= 1) || (ni >= na))
-                        if ni == 0
-                            C = A;
+        %SUBSASGN Subscripted assignment.
```
Indent method description, like you did for the reshape method

wollmich commented on @LinProp/LinProp.m:
```diff
@@ -431,161 +431,382 @@ function display(obj)
+               
+        end
+        function B = subsref(A, S)
+        %SUBSREF Subscripted reference.
```
Indent method description, like you did for the reshape method.

wollmich commented:
Can you merge the current main brunch into this brunch? So that `ndims` will work as expected.

wollmich commented on @MCProp/MCProp.m:
```diff
@@ -431,161 +431,390 @@ function display(obj)
-                    if ((ni <= 1) || (ni >= na))
-                        if ni == 0
-                            C = A;
+            %   A(I) = B assigns the values of B into the elements of A specified by
```
This line is missing:
%SUBSASGN Subscripted assignment.

wollmich replied:
By the way you can just edit the LinProp.m file, then you can open it an text editor replace all `LinProp` by `DistProp` or `MCProp` and save the file as DistProp.m or MCProp.m.

DionTimmermann commented:
The only error I am currenly still aware of is that `isempty(LinProp([]))` returns false. Do you consider this a separate issue or should it be fixed in this branch?

wollmich commented:
`isempty` is not implemented for *Prop. But it would be just a single line. So feel free to add it in this branch.

wollmich commented:
> This branch is 4 commits behind wollmich:main.

You should fetch the upstream of your main brunch.

wollmich commented:
Using MATLAB 2018a I get the following FAILED when using your unit_tests_subsasgn.m script:

```
>> compare_a_dbl_unc(rand(2, 3, 4), rand(3, 4), 'a(2, 1.1:3.1, :) = b;');
Warning: Integer operands are required for colon operator when used as index 
> In compare_a_dbl_unc (line 73)
 
FAILED: only unc caused an error. (Array indices must be positive integers or logical values.)
>> 
>> compare_a_dbl_unc([], magic(3), 'clear a; a(5, 1:9) = b;');
FAILED: different error messages. Expected 'Unable to perform assignment because the indices on the left side are not compatible with the size of the right side.', but got 'Unable to perform assignment because the size of the left side is 1-by-9 and the size of the right side is 3-by-3.'.
>> 
>> compare_a_dbl_unc([], magic(3), 'clear a; a(5, :) = b;');
FAILED: different error messages. Expected 'Unable to perform assignment because the indices on the left side are not compatible with the size of the right side.', but got 'Unable to perform assignment because the size of the left side is 1-by-9 and the size of the right side is 3-by-3.'.
>> 
```
This seems just to be different error messages. That's ok for me. Add a comment in your test script would help.

Using your unit_tests_subsref.m script I get several FAILED and one error:

```
Error using all
Dimension argument must be a positive integer scalar within indexing range.

Error in unit_tests_subsref (line 152)
if all(a_dbl(b) == double(a_unc(b)), 'all')
 
>> a_dbl(b) == double(a_unc(b))

  2×3×4 logical array

ans(:,:,1) =

   1   1   1
   1   1   1


ans(:,:,2) =

   1   1   1
   1   1   1


ans(:,:,3) =

   1   1   1
   1   1   1


ans(:,:,4) =

   1   1   1
   1   1   1

>> help all
 all    True if all elements of a vector are nonzero.
    For vectors, all(V) returns logical 1 (TRUE) if none of the elements 
    of the vector are zero.  Otherwise it returns logical 0 (FALSE).  For 
    matrices, all(X) operates on the columns of X, returning a row vector
    of logical 1's and 0's. For N-D arrays, all(X) operates on the first
    non-singleton dimension.
 
    all(X,DIM) works down the dimension DIM.  For example, all(X,1)
    works down the first dimension (the rows) of X.
 
    See also any.

    Reference page for all

>> 
```

and

```
>> compare_ans_dbl_unc(rand(5, 1), 'a([2 4 6 8])');
FAILED: different error messages. Expected 'Index exceeds array bounds.', but got 'Index exceeds the number of array elements (5).'.
>> 
```

The FAILED seems to be again just different error messages. That's ok for me. Add a comment to the test script.

DionTimmermann commented:
For me (MATLAB 21b) the following tests in https://github.com/DionTimmermann/metas-unclib-matlab-wrapper-tests also produce different error messages:
```
% Row vector in row of matrix
compare_a_dbl_unc(zeros(3, 3), [1, 2, 3, 4], 'a(1, :) = b;');
compare_a_dbl_unc(zeros(3, 3), [1, 2, 3, 4], 'a(2, :) = b;');
compare_a_dbl_unc(zeros(3, 3), [1, 2, 3, 4], 'a(3, :) = b;');
compare_a_dbl_unc(zeros(3, 3), [1, 2, 3, 4], 'a(4, :) = b;');
compare_a_dbl_unc(zeros(3, 3), [1, 2, 3, 4], 'a(5, :) = b;');

% Column vector in column of matrix
compare_a_dbl_unc(zeros(3, 3), [1; 2; 3; 4], 'a(:, 1) = b;');
compare_a_dbl_unc(zeros(3, 3), [1; 2; 3; 4], 'a(:, 2) = b;');
compare_a_dbl_unc(zeros(3, 3), [1; 2; 3; 4], 'a(:, 3) = b;');
compare_a_dbl_unc(zeros(3, 3), [1; 2; 3; 4], 'a(:, 4) = b;');
compare_a_dbl_unc(zeros(3, 3), [1; 2; 3; 4], 'a(:, 5) = b;');
```

These different error messages (the ones mentioned by @wollmich and the ones in this post) hint at differences in subsasgns and subsref for unc variables and double variables. These differences might result in different behavior in cases that have not been tested yet. However, the error messages with unc variables are more precise and the implementations for double variables are written in a different programming language. Therefore, I agree that these differences are acceptable and have added respective comments in the -tests repository.

I am currently not aware of any remaining issues.

wollmich commented:
Using MATLAB 2018a I get the following FAILED which are not commented as accepted when using your unit_tests_subsasgn.m script:

```
compare_a_dbl_unc(magic(3), [], 'a(:) = b;'); 
compare_a_dbl_unc(magic(3), [], 'a(2:3) = b;');
compare_a_dbl_unc(rand(3, 5, 7), [], 'a(2:3, :, :) = b;');
compare_a_dbl_unc(rand(2, 3, 4), [], 'a(:, 2, :) = b;');
compare_a_dbl_unc(rand(2, 3, 4), [], 'a(4, 2, :) = b;');
```

and the output:

```
FAILED: only double caused an error. (Unable to perform assignment because the left and right sides have a different number of elements.)
FAILED: only double caused an error. (Unable to perform assignment because the left and right sides have a different number of elements.)
FAILED: only double caused an error. (Unable to perform assignment because the size of the left side is 2-by-5-by-7 and the size of the right side is 0-by-0.)
FAILED: only double caused an error. (Unable to perform assignment because the size of the left side is 2-by-1-by-4 and the size of the right side is 0-by-0.)
FAILED: different error messages. Expected 'Unable to perform assignment because the size of the left side is 1-by-1-by-4 and the size of the right side is 0-by-0.', but got 'A null assignment can have only one non-colon index.'.
```

I've try to understand it:

```
>> a = magic(3);
>> b = [];
>> a(:) = b
Unable to perform assignment because the left and right sides have a different number
of elements.
 
>> a_unc = unc(a);
>> b_unc = unc(b);
>> a_unc(:) = b_unc

a_unc =

     []
```

wollmich commented:
Using MATLAB 2018a I get the following FAILED which are not commented as accepted when using your unit_tests_subsref.m script:

```
compare_ans_dbl_unc(rand(5, 1), 'a([2 4 6])'); 
compare_ans_dbl_unc(rand(5, 1), 'a([2 4 6 8])');
compare_ans_dbl_unc(rand(5, 1), 'a(10)'); 

compare_ans_dbl_unc(rand(4, 5), 'a(1:21)');
compare_ans_dbl_unc(rand(4, 5), 'a(21:23)');

compare_ans_dbl_unc(rand(4, 5), 'a(23)');

compare_ans_dbl_unc(rand(4, 5), 'a([1 3 3 4 1 6 8 30 4])');

compare_ans_dbl_unc([], 'a(4)');

compare_ans_dbl_unc(rand(2, 3, 4), 'a(''d'')');
```

Output:

```
FAILED: different error messages. Expected 'Index exceeds array bounds.', but got 'Index exceeds the number of array elements (5).'.
FAILED: different error messages. Expected 'Index exceeds array bounds.', but got 'Index exceeds the number of array elements (5).'.
FAILED: different error messages. Expected 'Index exceeds array bounds.', but got 'Index exceeds the number of array elements (5).'.

FAILED: different error messages. Expected 'Index exceeds array bounds.', but got 'Index exceeds the number of array elements (20).'.
FAILED: different error messages. Expected 'Index exceeds array bounds.', but got 'Index exceeds the number of array elements (20).'.

FAILED: different error messages. Expected 'Index exceeds array bounds.', but got 'Index exceeds the number of array elements (20).'.

FAILED: different error messages. Expected 'Index exceeds array bounds.', but got 'Index exceeds the number of array elements (20).'.

FAILED: different error messages. Expected 'Index exceeds array bounds.', but got 'Index exceeds the number of array elements (0).'.

FAILED: different error messages. Expected 'Index exceeds array bounds.', but got 'Index exceeds the number of array elements (24).'.
```

The error message of UncLib seems more precise, that's ok for me but should be marked as comment.

Perhaps a different error message should be a WARNING not a FAILED.

DionTimmermann commented:
I was able to reproduce and fix the FAILED tests caused by
```
compare_a_dbl_unc(magic(3), [], 'a(:) = b;'); 
compare_a_dbl_unc(magic(3), [], 'a(2:3) = b;');
compare_a_dbl_unc(rand(3, 5, 7), [], 'a(2:3, :, :) = b;');
compare_a_dbl_unc(rand(2, 3, 4), [], 'a(:, 2, :) = b;');
compare_a_dbl_unc(rand(2, 3, 4), [], 'a(4, 2, :) = b;');
```
I simly did not notice these failed tests. In case of a null assignment `a(x) = [];`, subsasgin is called with variable B being an empty double. I used the test isempyt(B) to test for this null assignment. Since isempty(Linprop([])) now returns true, subsasgin incorrectly assumed `a(x) = LinProp([])` would be a null assignment. I corrected this behavior in my last commit.

However I believe that a matlab implementation of subsasgn will never be able to differentiate between `a(x) = []`
 and `a(x) = double([])`. I added auch tests to my test script to document this difference and marked these new FAILED tests as accepted.

DionTimmermann commented:
I am not able to reproduce the failed these failed tests im MATLAB 2021b
```
compare_ans_dbl_unc(rand(5, 1), 'a([2 4 6])'); 
compare_ans_dbl_unc(rand(5, 1), 'a([2 4 6 8])');
compare_ans_dbl_unc(rand(5, 1), 'a(10)'); 

compare_ans_dbl_unc(rand(4, 5), 'a(1:21)');
compare_ans_dbl_unc(rand(4, 5), 'a(21:23)');

compare_ans_dbl_unc(rand(4, 5), 'a(23)');

compare_ans_dbl_unc(rand(4, 5), 'a([1 3 3 4 1 6 8 30 4])');

compare_ans_dbl_unc([], 'a(4)');

compare_ans_dbl_unc(rand(2, 3, 4), 'a(''d'')');
```
Probably a change in matalb that made the error message more precise.

In my test scripts, I will change the message for different errors from "FAILED" to "WARNING".

wollmich commented:
Ok now I think I'm ready to merge your pull request. Do you've anything to add?

Thanks a lot for your improvement of METAS UncLib MATLAB Wrapper.

If you want you can add yourself to the header:

```
% Metas.UncLib.Matlab.LinProp V2.4.8
% Michael Wollensack METAS - 28.05.2021
% Dion Timmermann PTB - 04.06.2021
%
```

DionTimmermann commented:
At the moment I have nothing to add. If anything else comes up, I believe it best to create an issue (and possibly propose a solution in a new branch/pull request).

Thank you for accepting this pull request! I think I would have gone mad, if I had not been able to use subsasgn for multiple elements at once ;-)
